data_layout.go

```diff
@@ -0,0 +1,421 @@
+}
+
+type PointerSizeAlignment struct { // All sizes are in bits.
+	addressSpace            uint64
+	size                    uint64
+	abiAlignment            uint64
+	preferredAlignment      uint64 // optional and defaults to abi
+	addressCalculationIndex uint64 // default 0
+}
+
+func NewPointerSizeAlignment(addSp, size, abiAl, prefAl, addCalInd uint64) *PointerSizeAlignment {
+	return &PointerSizeAlignment{addressSpace: addSp, size: size, abiAlignment: abiAl, preferredAlignment: prefAl, addressCalculationIndex: addCalInd}
+}
+
+func getDefaultPtrAligns() map[uint64]*PointerSizeAlignment {
+	ptrAlignMap := make(map[uint64]*PointerSizeAlignment)
+	ptrAlignMap[0] = NewPointerSizeAlignment(0, 64, 64, 64, 64)
+	return ptrAlignMap
+}
+
+type IntegerSizeAlignment struct { // All sizes are in bits.
+	size               uint64
+	abiAlignment       uint64
+	preferredAlignment uint64 // optional and defaults to abi
+}
+
+func NewIntegerSizeAlignment(size, abiAl, prefAl uint64) *IntegerSizeAlignment {
+	return &IntegerSizeAlignment{size: size, abiAlignment: abiAl, preferredAlignment: prefAl}
+}
+
+func getDefaultIntAligns() map[uint64]*IntegerSizeAlignment {
+	intAlignMap := make(map[uint64]*IntegerSizeAlignment)
+	intAlignMap[1] = NewIntegerSizeAlignment(1, 8, 8)
+	intAlignMap[8] = NewIntegerSizeAlignment(8, 8, 8)
+	intAlignMap[16] = NewIntegerSizeAlignment(16, 16, 16)
+	intAlignMap[32] = NewIntegerSizeAlignment(32, 32, 32)
+	intAlignMap[64] = NewIntegerSizeAlignment(64, 32, 64)
+	return intAlignMap
+}
+
+type VectorSizeAlignment struct { // All sizes are in bits.
+	size               uint64
+	abiAlignment       uint64
+	preferredAlignment uint64 // optional and defaults to abi
+}
+
+func NewVectorSizeAlignment(size, abiAl, prefAl uint64) *VectorSizeAlignment {
+	return &VectorSizeAlignment{size: size, abiAlignment: abiAl, preferredAlignment: prefAl}
+}
+
+func getDefaultVectorAligns() map[uint64]*VectorSizeAlignment {
+	vectorAlignMap := make(map[uint64]*VectorSizeAlignment)
+	vectorAlignMap[64] = NewVectorSizeAlignment(64, 64, 64)
+	vectorAlignMap[128] = NewVectorSizeAlignment(128, 128, 128)
+	return vectorAlignMap
+}
+
+type FloatingPointSizeAlignment struct { // All sizes are in bits.
+	size               uint64
+	abiAlignment       uint64
+	preferredAlignment uint64 // optional and defaults to abi
+}
+
+func NewFloatingPointSizeAlignment(size, abiAl, prefAl uint64) *FloatingPointSizeAlignment {
+	return &FloatingPointSizeAlignment{size: size, abiAlignment: abiAl, preferredAlignment: prefAl}
+}
+
+func getDefaultFloatAligns() map[uint64]*FloatingPointSizeAlignment {
+	floatAlignMap := make(map[uint64]*FloatingPointSizeAlignment)
+	floatAlignMap[16] = NewFloatingPointSizeAlignment(16, 16, 16)
+	floatAlignMap[32] = NewFloatingPointSizeAlignment(32, 32, 32)
+	floatAlignMap[64] = NewFloatingPointSizeAlignment(64, 64, 64)
+	floatAlignMap[128] = NewFloatingPointSizeAlignment(128, 128, 128)
+	return floatAlignMap
+}
+
+type AggregateAlignment struct {
```

Comment:
If possible, add a simple comment for each exported type to describe what it is used for. If there are documentation in the LLVM language reference, then it would be great to add links to those. If not, that's fine.

Reply:
I already added a common comment to the DataLayout struct with the Ref Doc link which has all the details. Since there are no anchor links to each spec property. I don't think a short text would do justice for the information given there and I'll end up copying whole description of each spec here. So I am thinking I'll leave it as it is. If you still prefer adding it, I'll do it but we'll need to track everytime a release happens for changes in main doc and update here appropriately.